Paper Implementation Chellenge #811

king-p3nguin opened the pull request:
### PR Description

Fix #811 

The added notebook explains the implementation of energy gradient calculation based on the BPDE algorithm using Classiq.
The notebook includes:
- Implementation of BPDE algorithm using Classiq library
- Demonstration of energy gradient calculation based on BPDE algorithm

I have attempted to reproduce the figures in the paper, but due to the high computational cost, I have only run the energy gradient calculation at a single point.

### Some notes

- [x] Please make sure that you placed the files in an appropriate folder
- [x] And that the files have indicative names.

- [x] Please note that Classiq runs automatic code linting, which may minorly alter some files.
  - [x] If you're familiar with `pre-commit`, you may run `pre-commit install`, and then at each commit, your files will be altered in a similar way

king-p3nguin commented:
I added auto-generated qmod and json files to follow the contribution guidelines.

orsa-classiq commented on community/paper_implementation_project/quantum_algorithm_for_numerical_energy_gradient_calculations_at_the_full_configuration_interaction_level_of_theory/energy_gradient.ipynb:
```diff
@@ -0,0 +1,745 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #21.                backend_name=ClassiqSimulatorBackendNames.SIMULATOR_STATEVECTOR
</pre><p>I think using the statevector is not so beneficial. You can just use a regular simulator and count the times you get |0&gt;, or use execution_session.estimate with pauli Z.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/910/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

orsa-classiq replied:
<pre class="ql-syntax" spellcheck="false">Line #4.        qmod = get_qfunc(
</pre><p>see my comment above about saving the re-compilation</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/910/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

king-p3nguin commented:
@orsa-classiq, Thank you for the review!
I am sorry for not getting back to you sooner. I made the change as requested.

Thank you for telling me about ExecutionSession! Now that the code execution time is faster, I have modified the last gradient calculation code to calculate a more accurate gradient. Also, the error bar was displayed incorrectly, so I also fixed that.

orsa-classiq commented on community/paper_implementation_project/bpde/bpde.metadata.json:
```diff
@@ -0,0 +1,11 @@
+  "description": "Bayesian Phase Difference Estimation Algorithm for Calculating Numerical Energy Gradient",
+  "problem_domain_tags": [],
+  "qmod_type": [
+    "community"
```
change to `Algorithms`

orsa-classiq commented on tests/resources/timeouts.yaml:
```diff
@@ -334,4 +334,14 @@ whats_classiq.qmod: 1000
\ No newline at end of file
+quantum_compression_algorithm_for_symmetric_states.qmod: 60
+bpde.ipynb: 1200
+bpde_test.qmod: 20
```
remove and rename the necessary files according to the comments in the notebook

orsa-classiq commented on tests/notebooks/test_bpde.py:
```diff
@@ -0,0 +1,32 @@
+)
+
+
+@wrap_testbook("bpde", timeout_seconds=1200)
```
this is very long. should be no longer than ~10 minutes. can you make it shorter?

orsa-classiq reviewed (changes requested):
@king-p3nguin Amazing work! the notebook looks very good :)
I put a few cosmetic comments. after you fix it it will be ready for merge :)

king-p3nguin commented:
@orsa-classiq 
Thank you for the comments! I made the change as requested.

orsa-classiq commented on community/paper_implementation_project/bpde/bpde.ipynb:
```diff
@@ -0,0 +1,820 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #20.        out_file=f"bpde_suzuki_trotter_reps_{int(np.ceil(time / 0.5))}",
</pre><p>please call this one just <code>bpde</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/910/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

orsa-classiq replied:
<pre class="ql-syntax" spellcheck="false">Line #20.            out_file=f"bpde_suzuki_trotter_reps_{int(np.ceil(time / 0.5))}",
</pre><p>please don't save any more qmods. so remove the line <code>out_file=f"bpda_su...</code></p><p> <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/910/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

orsa-classiq commented:
@king-p3nguin You didn't understand my comment about the qmods. We usualy save only a single qmod as a representative of the notebook. So please use the `out_file` argument only for the first example qmod you create (make sure to call it just `bpde`). remove the unecessary qmods from the commit. 
After that we will be able to merge

king-p3nguin commented:
@orsa-classiq 
Thank you for the clarification. I removed unnecessary `.qmod` files.

orsa-classiq commented:
@king-p3nguin great! everything is good now:)
it seems that there are conflicts, I think because you used the command `merge` from main, while we allow only rebase.
I suggest you open a new PR from a recent version of main, together with the relevant files. Please be careful not to use `merge` but only `rebase` if necessary. Please send me here the new PR link and we will merge it

king-p3nguin commented:
@orsa-classiq 
Thank you for the guidance! Here is the new PR just created: https://github.com/Classiq/classiq-library/pull/965
